src/qengine/state/state.cpp

```diff
@@ -87,28 +87,47 @@ void QEngineCPU::SetQuantumState(Complex16* inputState)
  *
  * A fundamental operation used by almost all gates.
  */
-void QEngineCPU::Apply2x2(bitCapInt offset1, bitCapInt offset2, const Complex16* mtrx, const bitLenInt bitCount,
+void QEngineCPU::Apply2x2(bitCapInt offset1, bitCapInt offset2, const complex* mtrx, const bitLenInt bitCount,
     const bitCapInt* qPowersSorted, bool doCalcNorm)
 {
-    Complex16 nrm = Complex16((bitCount == 1) ? (1.0 / runningNorm) : 1.0, 0.0);
     int numCores = GetConcurrencyLevel();
+    double nrm = 1.0 / runningNorm; 
+#if ENABLE_AVX  
+    ComplexUnion mtrxCol1;
+    ComplexUnion mtrxCol2;
+    mtrxCol1.cmplx[0] = mtrx[0];
+    mtrxCol1.cmplx[1] = mtrx[2];
+    mtrxCol2.cmplx[0] = mtrx[1];
+    mtrxCol2.cmplx[1] = mtrx[3];
+#else
+    complex nrm = complex(1.0 / runningNorm, 0.0);
+#endif
 
     if (doCalcNorm && (bitCount == 1)) {
         double* rngNrm = new double[numCores]; 
         std::fill(rngNrm, rngNrm + numCores, 0.0);
         par_for_mask(0, maxQPower, qPowersSorted, bitCount, [&](const bitCapInt lcv, const int cpu) {
-            Complex16 qubit[2];
+#if ENABLE_AVX
```

Comment:
Either fork Apply2x2 entirely - probably my recommendation, given how much of the code is non-duplicated - or place the #if and #endif outside the par_for_mask (so that each branch has a complete clause within it).  It’ll read cleaner as a result.

Reply:
Forked entirely, with the complex2 code.